lib/pbench/server/cache_manager.py

```diff
@@ -1433,3 +1505,124 @@ def delete(self, dataset_id: str):
         del self.tarballs[name]
 
         self._clean_empties(tarball.controller_name)
+
+    def reclaim_cache(self, goal_pct: float = 20.0, goal_bytes: int = 0) -> bool:
+        """Reclaim unused caches until at least goal (%) disk is available.
+
+        The cache tree need not be fully discovered at this point; we can
+        reclaim cache even on a partial tree. This is driven by discovery of
+        the cache directory tree, looking for <resource_id> directories that
+        contain an unpacked tarball root rather than only the `lock` and
+        `last_ref` files.
+
+        This is a "best effort" operation. It will free least recently used
+        caches that are currently unlocked until both the goal % is achieved
+        and the absolute goal bytes is achieved (if specified), or until there
+        are no more unlocked cached tarballs.
+
+        Args:
+            tree: the cache manager instance
+            goal_pct: goal percent of cache filesystem free
+            goal_bytes: goal in bytes
+
+        Returns:
+            True if both goals are met, or False otherwise
+        """
+
+        @dataclass
+        class Candidate:
+            last_ref: float  # last reference timestamp
+            cache: Path  # Unpacked artifact directory
+
+        def reached_goal():
+            usage = shutil.disk_usage(self.cache_root)
+            available = float(usage.free) / float(usage.total) * 100.0
+            return bool(available >= goal_pct and usage.free >= goal_bytes)
```

Comment:
It's not "redundant"; and while it's certainly possible to consolidate them, there wasn't that much logic and it seemed clearer/cleaner this way.

Now, of course, you've got it stuck in my brain ... 😦

Reply:
It strikes me that offering the user two ways to express the goal (when there is only one way to satisfy it...) is a nice convenience for the user (although, permitting the user to express "the goal" in two ways _simultaneously_ (i.e., accepting _both_ the percentage **and** the total bytes numbers) is kind of letting the user be lazy...), but this all amounts to a "user interface thing" -- that is, once command line processing is done, the result should really be a single number, and that number should be bytes, (because we're going to compare it to `usage.free`).

So, we should really be calculating `goal_bytes` in `main()` and the code here should devolve to a simple comparison.

Reply:
That's certainly one possible definition of the logic; but it requires that the UI know the drive we're targeting and understand how to get and process the usage rather than isolating that logic within the cache manager as I've done.

Whether the %/bytes goals are consolidated here or maintained separately seems like a pretty minor issue.

Reply:
Your comment about isolation is an excellent one.  But, I still think the consolidation would best be done in the UI processing code.  I probably would have solved that by having the UI code call a subroutine in the cache manager.

I agree that whether the goals are consolidated or maintained separately _is_ a pretty minor issue, but that doesn't mean it's not better to consolidate them.  🙂